compiler-core/src/language_server/annotation_printer.rs

```diff
@@ -0,0 +1,218 @@
+#![allow(dead_code)]
+
+use ecow::EcoString;
+use im::HashMap;
+use itertools::Itertools;
+use std::sync::Arc;
+
+use crate::{
+    ast::TypeAst,
+    type_::{Type, TypeVar, PRELUDE_MODULE_NAME},
+};
+
+#[derive(Debug)]
+pub struct AnnotationPrinter<'a> {
+    names: HashMap<u64, EcoString>,
+    uid: u64,
+
+    // A mapping of printd type names to the module that they are defined in.
+    printed_types: HashMap<EcoString, EcoString>,
+
+    // Mapping of module.type to type alias -
+    // eg: import mod1.{type Cat as c} -> Key: (mod1, Cat), Value: c
+    // eg: type UserId = Int -> Key: (mod, UserId), Value: Int
+    type_aliases: &'a HashMap<(EcoString, EcoString), EcoString>,
+
+    // Mapping of module alias to module name
+    module_aliases: &'a HashMap<EcoString, EcoString>,
+    type_parameters: &'a HashMap<u64, TypeAst>,
+    current_module: EcoString,
+}
+
+impl<'a> AnnotationPrinter<'a> {
+    pub fn new(
+        type_aliases: &'a HashMap<(EcoString, EcoString), EcoString>,
+        module_aliases: &'a HashMap<EcoString, EcoString>,
+        type_parameters: &'a HashMap<u64, TypeAst>,
+        current_module: EcoString,
+    ) -> Self {
+        AnnotationPrinter {
+            names: im::HashMap::new(),
+            uid: u64::default(),
+            printed_types: HashMap::new(),
+            type_aliases,
+            module_aliases,
+            type_parameters,
+            current_module,
+        }
+    }
+
+    pub fn print(&mut self, typ: &Type) -> String {
```

Comment:
This method constructs a great many strings- instead let's have it push strs onto the mutable buffer so it only ever allocates one and then grows that as needed.

I'd do this by having a public method that returns a string, and internally it resets an buffer and then calls a private printing function which recursively calls itself and each call will append to the buffer.

Reply:
Gotcha Thank you !